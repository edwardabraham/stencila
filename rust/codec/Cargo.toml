[package]
name = "codec"
description = "Defines a specification and trait for codecs"
version = "0.0.0"
edition = "2021"

[dependencies]
common = { path = "../common" }
stencila-schema = { path = "../schema" }
utils = { path = "../utils" }
