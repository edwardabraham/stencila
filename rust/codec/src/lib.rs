use std::{fs::create_dir_all, path::Path};

use common::{
    async_trait::async_trait,
    eyre::{bail, Result},
    serde::{Deserialize, Serialize},
    tokio::{
        fs::File,
        io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt, BufReader, BufWriter},
    },
};
use stencila_schema::Node;
use utils::vec_string;

// Re-export for the convenience of crates that implement `CodecTrait`
pub use common;
pub use stencila_schema;
pub use utils;

/// A specification for codecs
///
/// All codecs, including those implemented in plugins, should provide this
/// specification. Rust implementations return a `Codec` instance from the
/// `spec` function of `CodecTrait`. Plugins provide a JSON or YAML serialization
/// as part of their manifest.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(default, rename_all = "camelCase", crate = "common::serde")]
pub struct Codec {
    /// A list of format names (or aliases) that the codec can decode from
    /// or encode to
    pub formats: Vec<String>,

    /// Whether the codec supports decoding from string content
    pub from_string: bool,

    /// Whether the codec supports decoding from a file system path
    pub from_path: bool,

    /// Whether the codec supports encoding to string content
    pub to_string: bool,

    /// Whether the codec supports encoding to a file system path
    pub to_path: bool,

    /// A list of root node types that the codec can encode / decode
    ///
    /// Most codecs usually only handle one root type e.g. `Article`.
    /// Used to provide a list of formats to the user that support
    /// the current document type.
    pub root_types: Vec<String>,

    /// A list of node types that the codec does not support
    ///
    /// Used to provide warnings to the user on potential loss of content
    /// when encoding using the codec.
    pub unsupported_types: Vec<String>,

    /// A list of node properties that the codec does not support
    ///
    /// The format for these strings is `<type>.<property>` e.g. `Article.funders`
    /// Used to provide warnings to the user on potential loss of content
    /// when encoding using the codec.
    pub unsupported_properties: Vec<String>,

    /// The status of the codec e.g. `alpha`, `beta`, `stable`
    pub status: String,
}

impl Default for Codec {
    fn default() -> Self {
        Self {
            formats: vec![],
            from_string: true,
            from_path: true,
            to_string: true,
            to_path: true,
            root_types: vec_string!["Article"],
            unsupported_types: vec![],
            unsupported_properties: vec![],
            status: "alpha".to_string(),
        }
    }
}

/// A trait for codecs
///
/// This trait can be used by Rust implementations of codecs, allowing them to
/// be compiled into the Stencila binaries.
///
/// It defines similar functions to `serde_json` (and other `serde_` crates) for
/// converting nodes to/from strings, files, readers etc.
#[async_trait]
pub trait CodecTrait {
    /// Get the [`Codec`] specification for this implementation
    fn spec() -> Codec;

    /// Decode a document node from a string
    fn from_str(_str: &str, _options: Option<DecodeOptions>) -> Result<Node> {
        bail!("Decoding from string is not implemented for this format")
    }

    /// Decode a document node from a string asynchronously
    async fn from_str_async(str: &str, options: Option<DecodeOptions>) -> Result<Node> {
        Self::from_str(str, options)
    }

    /// Decode a document node from a `BufReader`
    async fn from_buffer<T: AsyncRead>(
        reader: &mut BufReader<T>,
        options: Option<DecodeOptions>,
    ) -> Result<Node>
    where
        T: Send + Sync + Unpin,
    {
        let mut content = String::new();
        reader.read_to_string(&mut content).await?;
        Self::from_str_async(&content, options).await
    }

    /// Decode a document node from a file
    ///
    /// This function reads the file as a string and passes that on to `from_str`
    /// for decoding. If working with binary formats, you should override this function
    /// to read the file as bytes instead.
    async fn from_file(file: &mut File, options: Option<DecodeOptions>) -> Result<Node> {
        let mut content = String::new();
        file.read_to_string(&mut content).await?;
        Self::from_str_async(&content, options).await
    }

    /// Decode a document node from a file system path
    async fn from_path(path: &Path, options: Option<DecodeOptions>) -> Result<Node> {
        let mut file = File::open(path).await?;
        Self::from_file(&mut file, options).await
    }

    /// Encode a document node to a string
    fn to_string(_node: &Node, _options: Option<EncodeOptions>) -> Result<String> {
        bail!("Encoding to string is not implemented for this format")
    }

    /// Encode a document node to a string asynchronously
    async fn to_string_async(node: &Node, options: Option<EncodeOptions>) -> Result<String> {
        Self::to_string(node, options)
    }

    /// Encode a document node to a `BufWriter`
    async fn to_buffer<T: AsyncWrite>(
        node: &Node,
        writer: &mut BufWriter<T>,
        options: Option<EncodeOptions>,
    ) -> Result<()>
    where
        T: Send + Sync + Unpin,
    {
        let content = Self::to_string_async(node, options).await?;
        writer.write_all(content.as_bytes()).await?;
        Ok(())
    }

    /// Encode a document node to a file
    async fn to_file(node: &Node, file: &mut File, options: Option<EncodeOptions>) -> Result<()> {
        let content = Self::to_string_async(node, options).await?;
        file.write_all(content.as_bytes()).await?;
        Ok(())
    }

    /// Encode a document node to a file system path
    async fn to_path(node: &Node, path: &Path, options: Option<EncodeOptions>) -> Result<()> {
        if let Some(parent) = path.parent() {
            create_dir_all(parent)?
        }
        let mut file = File::create(path).await?;
        Self::to_file(node, &mut file, options).await
    }
}

/// Decoding options
///
/// Decoding functions (including those in plugins) are encouraged to respect these options
/// but are not required to.
#[derive(Clone, Default)]
pub struct DecodeOptions {
    /// The format of the input to be decoded
    ///
    /// Most codecs only decode one format. However, for those that handle multiple
    /// format it may be necessary to specify this option.
    pub format: Option<String>,
}

/// Encoding options
///
/// Encoding functions (including those in plugins) are encouraged to respect these options
/// but are not required to. Indeed, some options do not apply for some formats.
/// For example, a PDF is always `standalone` (so if that option is set to `false`, it will be ignored).
/// Futhermore, some combinations of options are ineffectual e.g. a `theme` when `standalone: false`
#[derive(Clone)]
pub struct EncodeOptions {
    /// Whether to encode in compact form
    ///
    /// Some formats (e.g HTML and JSON) can be encoded in either compact
    /// or "pretty-printed" (e.g. indented) forms.
    pub compact: bool,

    /// Whether to ensure that the encoded document is standalone
    ///
    /// Some formats (e.g. Markdown, DOCX) are always standalone.
    /// Others can be fragments, or standalone documents (e.g HTML).
    pub standalone: bool,

    /// Whether to bundle local media files into the encoded document
    ///
    /// Some formats (e.g. DOCX, PDF) always bundle. For HTML, bundling means
    /// including media as data URIs rather than links to files.
    pub bundle: bool,

    /// Whether to use Stencila Cloud as a remote storage for document nodes
    /// that can not be wholly represented within the format
    ///
    /// Some formats (e.g. Google Docs) are not able to fully represent some types
    /// of nodes (e.g. CodeChunks). In these cases we store the node on Stencila Cloud
    /// with a unique id and key and insert a link to it from within the format.
    pub rpng_content: bool,
    pub rpng_link: bool,

    /// The theme to apply to the encoded document
    ///
    /// Only applies to standalone HTML and formats derived from it (e.g. PDF, PNG).
    pub theme: Option<String>,

    /// Whether Javascript for Web Components should be loaded or not
    ///
    /// Only applies to standalone HTML and formats derived from it (e.g. PDF, PNG).
    pub components: bool,

    /// The format to encode to
    ///
    /// Most codecs only encode to one format. However, for those that handle multiple
    /// formats it may be necessary to specify this option.
    pub format: Option<String>,
}

impl Default for EncodeOptions {
    fn default() -> Self {
        Self {
            compact: true,
            standalone: false,
            bundle: false,
            rpng_content: false,
            rpng_link: false,
            theme: None,
            components: true,
            format: None,
        }
    }
}
