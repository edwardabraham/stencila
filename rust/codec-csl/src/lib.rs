pub use codec::CodecTrait;
use codec::{
    common::{
        eyre::{bail, eyre, Result},
        serde_json,
    },
    stencila_schema::{
        Article, CreativeWorkAuthors, CreativeWorkPublisher, CreativeWorkTypes, Date,
        InlineContent, Node, Organization, Periodical, Person, PublicationIssue,
        PublicationIssueIssueNumber, PublicationVolume, PublicationVolumeVolumeNumber,
        ThingIdentifiers,
    },
    utils::vec_string,
    Codec, DecodeOptions,
};

/// A codec for Citation Style Language (CSL) JSON
///
/// This uses `serde_json` to parse the JSON into a hash map and then extracts
/// various properties from it to populate an instance of one of the `CreativeWork` types,
/// for example, an `Article`, `Book` or `Dataset`.
///
/// The `citeproc-io` crate, which is a part of https://github.com/zotero/citeproc-rs, could be
/// used for parsing CSL-JSON:
///
///    let reference: citeproc_io::Reference = serde_json::from_str(str)?;
///
/// but it was found to be quite strict (e.g did not handle `"type": "journal-article"`)
/// and the parsed `Reference` still needed to be translated to a Stencila node.
pub struct CslCodec;

impl CslCodec {
    pub fn from_json(data: serde_json::Value) -> Result<Node> {
        let mut article = Article::default();

        if let Some(title) = data.get("title") {
            article.title = Some(vec![InlineContent::String(decode_string(title)?)]);
        }

        if let Some(date) = data.get("published") {
            article.date_published = Some(decode_date(date).map(Box::new)?);
        }
        if let Some(date) = data.get("submitted") {
            article.date_received = Some(decode_date(date).map(Box::new)?);
        }

        if let Some(authors) = data.get("author") {
            let authors = if let Some(authors) = authors.as_array() {
                authors
            } else {
                bail!("Expected `authors` to be an array")
            };
            let authors = authors
                .iter()
                .map(|author| {
                    let person = decode_person(author).unwrap();
                    CreativeWorkAuthors::Person(person)
                })
                .collect();
            article.authors = Some(authors);
        }

        let is_part_of = data
            .get("container-title")
            .and_then(|title| decode_string(title).ok())
            .map(|name| {
                Box::new(CreativeWorkTypes::Periodical(Periodical {
                    name: Some(Box::new(name)),
                    ..Default::default()
                }))
            });

        let is_part_of = if let Some(volume_number) = data.get("volume").and_then(|num| {
            num.as_str()
                .map(|num| PublicationVolumeVolumeNumber::String(num.to_string()))
                .or_else(|| num.as_i64().map(PublicationVolumeVolumeNumber::Integer))
        }) {
            Some(Box::new(CreativeWorkTypes::PublicationVolume(
                PublicationVolume {
                    is_part_of,
                    volume_number: Some(Box::new(volume_number)),
                    ..Default::default()
                },
            )))
        } else {
            is_part_of
        };

        let is_part_of = if let Some(issue_number) = data.get("volume").and_then(|num| {
            num.as_str()
                .map(|num| PublicationIssueIssueNumber::String(num.to_string()))
                .or_else(|| num.as_i64().map(PublicationIssueIssueNumber::Integer))
        }) {
            Some(Box::new(CreativeWorkTypes::PublicationIssue(
                PublicationIssue {
                    is_part_of,
                    issue_number: Some(Box::new(issue_number)),
                    ..Default::default()
                },
            )))
        } else {
            is_part_of
        };

        article.is_part_of = is_part_of;

        if let Some(publisher) = data.get("publisher") {
            let name = decode_string(publisher)?;
            article.publisher = Some(Box::new(CreativeWorkPublisher::Organization(
                Organization {
                    name: Some(Box::new(name)),
                    ..Default::default()
                },
            )));
        }

        Ok(Node::Article(article))
    }
}

impl CodecTrait for CslCodec {
    fn spec() -> Codec {
        Codec {
            formats: vec_string!["csl", "csl-json"],
            root_types: vec_string!["Article"],
            ..Default::default()
        }
    }

    fn from_str(str: &str, _options: Option<DecodeOptions>) -> Result<Node> {
        let data = serde_json::from_str(str)?;
        Self::from_json(data)
    }
}

/// Decode a `string` variable
fn decode_string(string: &serde_json::Value) -> Result<String> {
    string
        .as_str()
        .map(|str| str.to_string())
        .ok_or_else(|| eyre!("Expected a string value"))
}

/// Decode a `name-variable` into a `Person`
///
/// See https://github.com/citation-style-language/schema/blob/506040022f6c37846343edd36658b23e85b5b8ff/schemas/input/csl-data.json#L463
fn decode_person(data: &serde_json::Value) -> Result<Person> {
    let given_names = data
        .get("given")
        .and_then(|name| name.as_str())
        .map(|name| {
            name.split(' ')
                .map(|name| name.to_string())
                .collect::<Vec<String>>()
        });
    let family_names = data
        .get("family")
        .and_then(|name| name.as_str())
        .map(|name| {
            name.split(' ')
                .map(|name| name.to_string())
                .collect::<Vec<String>>()
        });
    let identifiers = data
        .get("ORCID")
        .or_else(|| data.get("orcid"))
        .and_then(|orcid| orcid.as_str())
        .map(|orcid| vec![ThingIdentifiers::String(orcid.to_string())]);

    Ok(Person {
        given_names,
        family_names,
        identifiers,
        ..Default::default()
    })
}

/// Decode a `date-variable` into a `Date`
///
/// See https://github.com/citation-style-language/schema/blob/506040022f6c37846343edd36658b23e85b5b8ff/schemas/input/csl-data.json#L499
fn decode_date(data: &serde_json::Value) -> Result<Date> {
    let value = if let Some(raw) = data.get("raw") {
        raw.as_str()
            .ok_or_else(|| eyre!("Expected `raw` to be a string"))?
            .to_string()
    } else if let Some(literal) = data.get("literal") {
        literal
            .as_str()
            .ok_or_else(|| eyre!("Expected `literal` to be a string"))?
            .to_string()
    } else if let Some(parts) = data.get("date-parts") {
        let parts = parts
            .as_array()
            .ok_or_else(|| eyre!("Expected `parts` to be an array"))?;
        let parts = parts
            .first()
            .and_then(|item| item.as_array())
            .ok_or_else(|| eyre!("Expected `parts` to have at least one item that is an array"))?;
        let mut date = parts
            .first()
            .and_then(|year| {
                year.as_str()
                    .map(|year| year.to_string())
                    .or_else(|| year.as_u64().map(|year| year.to_string()))
            })
            .ok_or_else(|| eyre!("Expected `year` to be a number"))?;
        if let Some(month) = parts.get(1).and_then(|month| {
            month
                .as_str()
                .map(|month| month.to_string())
                .or_else(|| month.as_u64().map(|month| month.to_string()))
        }) {
            date.push('-');
            date.push_str(&month);
        }
        if let Some(day) = parts.get(2).and_then(|day| {
            day.as_str()
                .map(|day| day.to_string())
                .or_else(|| day.as_u64().map(|day| day.to_string()))
        }) {
            date.push('-');
            date.push_str(&day);
        }
        date
    } else {
        bail!("Unable to parse `Date` from `{:?}`", data)
    };

    Ok(Date {
        value,
        ..Default::default()
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use test_snaps::{insta::assert_json_snapshot, snapshot_fixtures_content};

    #[test]
    fn fragments() {
        snapshot_fixtures_content("fragments/csl/*.json", |content| {
            let node = CslCodec::from_str(content, None).unwrap();
            assert_json_snapshot!(node);
        });
    }
}
