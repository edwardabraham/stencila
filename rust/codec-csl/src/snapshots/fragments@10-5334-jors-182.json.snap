---
source: rust/codec-csl/src/lib.rs
assertion_line: 237
expression: node
input_file: fixtures/fragments/csl/10-5334-jors-182.json

---
{
  "type": "Article",
  "authors": [
    {
      "type": "Person",
      "familyNames": [
        "Carlsson"
      ],
      "givenNames": [
        "Kristoffer"
      ],
      "identifiers": [
        "http://orcid.org/0000-0001-9092-3092"
      ]
    },
    {
      "type": "Person",
      "familyNames": [
        "Ekre"
      ],
      "givenNames": [
        "Fredrik"
      ],
      "identifiers": [
        "http://orcid.org/0000-0003-2476-5406"
      ]
    }
  ],
  "datePublished": {
    "type": "Date",
    "value": "2019"
  },
  "isPartOf": {
    "type": "PublicationIssue",
    "isPartOf": {
      "type": "PublicationVolume",
      "isPartOf": {
        "type": "Periodical",
        "name": "Journal of Open Research Software"
      },
      "volumeNumber": "7"
    },
    "issueNumber": "7"
  },
  "publisher": {
    "type": "Organization",
    "name": "Ubiquity Press, Ltd."
  },
  "title": "Tensors.jl â€” Tensor Computations in Julia"
}
